### PR TITLE
WiP trainer implementation

### PR DESCRIPTION
**Pull request by XapaJIaMnu**
Working on the trainer, all together with example config and test files

### CODE REVIEW

**Comment by ZJaume** [comment]
Been doing a couple more runs. Right now the child marian process seems to be poorly controlled. If I press CTRL+C the process stays there.

File: trainer/trainer.py [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+                #print(epoch, dataset, stage.until_dataset, stage.until_epoch, len(lines))
+                batch.extend(lines)
+            # Shuffle the batch
+            random.shuffle(batch)
+            # Uppercase randomly
+            batch = [x.upper() if random.random() < self.uppercase_ratio else x for x in batch]
+            self.trainer.stdin.writelines(batch)
+            self.state_tracker.update_seed()
+            # Termination condition, check if we finished training.
+            stop_training = self.dataset_objects[stage.until_dataset].epoch >= stage.until_epoch
```
Comment (ZJaume) [inline trainer/trainer.py:197]: Maybe we need a `self.trainer.wait()` somewhere here? The `communicate` function takes complete control of the suprocess but it seems to not fit the purpose of this function.

Comment (ZJaume) [reply trainer/trainer.py:197]: I think the wait should avoid the trainer.py to end before marian won't kill marian if CTRl+C. Maybe try to catch the CTRL+C exception and then wait()?

Comment (ZJaume) [reply trainer/trainer.py:197]: I think it would be:
```python
try:
     # train loop
except KeyboardInterrupt:
     trainer.wait()
     sys.exit()
```

Comment (jelmervdl) [reply trainer/trainer.py:197]: Add a `trainer.stdin.close()` before that `trainer.wait()` and that should stop the trainer cleanly. Alternatively you can use `trainer.kill()` to terminate it immediately. 

What I often do is:
```py
try:
  # train loop
finally:
  trainer.stdin.close()
  trainer.wait()
```

That way, in both the interrupted and the normal case you end with stopping the trainer cleanly. Although in this use case I'm not sure whether `trainer.kill()` might be helpful in an except so you don't have to wait too long.

Comment (XapaJIaMnu) [reply trainer/trainer.py:202]: Should be fixed

File: trainer/trainer.py [hunk 2]
```diff
@@ -109,6 +109,7 @@ def __init__(self, ymlpath: str, tmpdir: str, state_tracker: StateTracker):
             self.dataset_paths[path.split('/')[-1]] = path
         self.stage_names: List[str] = ymldata['stages']
         self.uppercase_ratio = float(ymldata['uppercase'])
+        self.tittlecase_ratio = float(ymldata['tittlecase'])
```
Comment (ZJaume) [inline trainer/trainer.py:112]: `tittlecase` typo?

**Comment by XapaJIaMnu** [comment]
Close, superseded by #52

**Pull request merged**